mailet/icalendar/src/main/java/org/apache/james/transport/mailets/model/ICALAttributeDTO.java

```diff
@@ -32,54 +32,50 @@
             VEvent vevent = (VEvent) calendar.getComponent("VEVENT");
-            this.uid = optionalOf(vevent.getUid());
-            this.method = optionalOf(calendar.getMethod());
-            this.recurrenceId = optionalOf(vevent.getRecurrenceId());
-            this.sequence = optionalOf(vevent.getSequence());
-            this.dtstamp = optionalOf(vevent.getDateStamp());
-            return this;
+            Optional<String> uid = optionalOf(vevent.getUid());
+            Optional<String> method = optionalOf(calendar.getMethod());
+            Optional<String> recurrenceId = optionalOf(vevent.getRecurrenceId());
+            Optional<String> sequence = optionalOf(vevent.getSequence());
+            Optional<String> dtstamp = optionalOf(vevent.getDateStamp());
+            Preconditions.checkNotNull(ical);
+            Preconditions.checkState(uid.isPresent(), "uid is a compulsory property of an ICAL object");
+            Preconditions.checkState(method.isPresent(), "method is a compulsory property of an ICAL object");
+            Preconditions.checkState(dtstamp.isPresent(), "dtstamp is a compulsory property of an ICAL object");
+            return sender -> recipient -> replyTo ->
+                    new ICALAttributeDTO(
+                            ical,
+                            uid.get(), sender.asString(),
+                            recipient.asString(),
+                            replyTo.asString(),
+                            dtstamp.get(), method.get(), sequence,
+                            recurrenceId);
         }
 
         private Optional<String> optionalOf(Property property) {
             return Optional.ofNullable(property).map(Property::getValue);
         }
 
-        public Builder sender(String sender) {
-            this.sender = sender;
-            return this;
+        @FunctionalInterface
+        public interface RequiresSender {
+            RequiresRecipient sender(MailAddress sender);
         }
 
-
-        public Builder recipient(MailAddress recipient) {
-            this.recipient = recipient.asString();
-            return this;
+        @FunctionalInterface
+        public interface RequiresRecipient {
+            RequiresReplyTo recipient(MailAddress recipient);
         }
 
-        public ICAL build() {
-            Preconditions.checkNotNull(recipient);
-            Preconditions.checkNotNull(sender);
-            Preconditions.checkNotNull(ical);
-            Preconditions.checkState(uid.isPresent(), "uid is a compulsary property of an ICAL object");
-            Preconditions.checkState(method.isPresent(), "method is a compulsary property of an ICAL object");
-            Preconditions.checkState(dtstamp.isPresent(), "dtstamp is a compulsary property of an ICAL object");
-            return new ICAL(ical, sender, recipient, uid, sequence, dtstamp, method, recurrenceId);
+        @FunctionalInterface
+        public interface RequiresReplyTo {
```

Comment:
why do you make it mandatory?

I would have thought it would have be sufficient (and more correct) to map the "replyTo" header to "replyTo" JSON attribute. If header is present, put the attribute, else not.

Or did I miss something?

Reply:
From the consumer of this change point of view, `replyTo` attribute is not mandatory. If `replyTo` is not presented, the `sender` will be used.

Reply:
If replyTo is null then OpenPaaS can make the choice to default to the sender. We should not enforce such choice in James.

Reply:
Let's turn it around:
 * When `replyTo` is missing `sender` is chosen (ie. specifying it anyway will not change the result)
 * When a `replyTo` is passed in the header and forget in the JSON, the `sender` is selected and is wrong

This design add some usually "extra work", but prevents tricky errors, I think the benefits of it largely outdo drawbacks.

Reply:
ok you are explaining it here. If @tuanlc agrees, I'm ok with this

Reply:
Yes, I totally agree with this. Thank @blackheaven for the explanation!